### PR TITLE
Adding Support for Private DNS Zones

### PR DESCRIPTION
**Pull request by jaiveerk**
PR to deploy modified external DNS resources for private DNS zones, plus testing suite that supports private clusters/private DNS zones.

Annotation changes came from this tutorial from ExternalDNS: https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/azure-private-dns.md

### CODE REVIEW

File: .github/workflows/e2e-private-cluster.yaml
Comment (OliverMKing) [inline .github/workflows/e2e-private-cluster.yaml:16]: checkout version is outdated. Bump the versions on these

File: .idea/workspace.xml
Comment (OliverMKing) [inline .idea/workspace.xml]: this file shouldn't be committed

Comment (OliverMKing) [reply .idea/workspace.xml]: you could add it to the gitignore

File: pkg/config/config.go
Comment (OliverMKing) [inline pkg/config/config.go:24]: Can we detect if a user's dns zone is private instead of adding a flag here? 

Are the rp code changes ready to support this flag ready?

File: devenv/private-cluster.yaml
Comment (OliverMKing) [inline devenv/private-cluster.yaml:1]: It seems incorrect to test this way here. We should be relying on the exact k8s manifests that rp uses to guarantee it, not creating arbitrary manifests that are meaningless in the context of the real application.

File: .github/workflows/e2e-private-cluster.yaml
Comment (OliverMKing) [inline .github/workflows/e2e-private-cluster.yaml:3]: do we want to set this up on pr? We can make it so a member of our team has to be the one to run

File: devenv/scripts/deploy_private_e2e_tester.sh
Comment (OliverMKing) [inline devenv/scripts/deploy_private_e2e_tester.sh:31]: why are we using this script instead of using go tests like the other e2e tests?

File: Makefile
Comment (davidgamero) [inline Makefile:24]: switching to /usr/bin/env instead of /bin/sh would make it more compatible with other shells users could have when develping

File: pkg/manifests/manifests.go
Comment (davidgamero) [inline pkg/manifests/manifests.go:247]: extracting several of the magic strings from the Private and Normal version of this function will reduce opportunities for drift

**Review by davidgamero — DISMISSED** [review]

File: .idea/workspace.xml
Comment (jaiveerk) [reply .idea/workspace.xml]: hm I've tried adding it multiple times with git rm, not sure why it's being so persistent. Got it removed earlier but it keeps coming back. Will look into this.

File: pkg/manifests/manifests.go
Comment (OliverMKing) [inline pkg/manifests/manifests.go:248]: instead of creating a new object with 90% of it being the same as `newIngressControllerService`, just modify the object that is returned with that function. It looks like you're just adding annotations.

**Review by OliverMKing — CHANGES_REQUESTED** [review]
Left some comments

File: devenv/private-cluster.yaml
Comment (jaiveerk) [reply devenv/private-cluster.yaml:1]: that's a good point - I just took this from the existing e2e framework. I imagine it's a copy of what RP uses anyway, but I'll look to see what I can do to synchronize the two.

File: .github/workflows/e2e-private-cluster.yaml
Comment (jaiveerk) [reply .github/workflows/e2e-private-cluster.yaml:3]: I tried it but ran into the same issue where github secrets can't be accessed for PRs across forks, so az login fails on PRs

File: devenv/scripts/deploy_private_e2e_tester.sh
Comment (jaiveerk) [reply devenv/scripts/deploy_private_e2e_tester.sh:31]: this uses the same e2e test that is used for public clusters. This script simply runs that same go code on the private cluster by deploying a pod on the private cluster that then calls make e2e.
We can't do it via go code because we can't access private clusters via a Kubernetes client on a machine outside the private cluster's VM.
Plus, this way, we have one suite that runs for both private and public clusters, so if changes are made/tests are added to this one suite, those changes propagate to both kinds of clusters.

File: pkg/manifests/manifests.go
Comment (jaiveerk) [reply pkg/manifests/manifests.go:248]: yep! just adding annotations. wasn't sure which would be clearer, but can definitely switch to just modifying fields in the existing object if you think that makes more sense/is more readable.

File: pkg/manifests/manifests.go
Comment (jaiveerk) [reply pkg/manifests/manifests.go:247]: good call!

File: devenv/private-cluster.yaml
Comment (jaiveerk) [reply devenv/private-cluster.yaml:1]: I think the smartest way to go about this will just be to use some sort of embedded file system that we can import from RP, but given how many other changes we're planning for this add-on, I'm gonna regard that as out of scope and a test enhancement that we can add later as we start changing stuff up on the testing side. Just want to have SOME tests at the moment.

File: pkg/manifests/manifests.go
Comment (OliverMKing) [inline pkg/manifests/manifests.go:69]: this is another example of something that shouldn't be moved to the top. This only has context inside of the nginxIngressControllerDeployment. Moving it here provides no benefit and just makes it way harder to read the code

**Review by OliverMKing — CHANGES_REQUESTED** [review]

File: pkg/config/config.go
Comment (jaiveerk) [reply pkg/config/config.go:43]: the check happens in RP when users link the DNS zone to their add-on. We already perform resource parsing in RP at that step (to extract information like the DNS zone domain and subscription ID) so it makes the most sense to also check whether or not it's private at that step.

File: pkg/manifests/manifests.go
Comment (jaiveerk) [reply pkg/manifests/manifests.go:32]: I'll re-introduce some of the const values as regular strings

File: pkg/manifests/manifests.go
Comment (jaiveerk) [reply pkg/manifests/manifests.go:69]: makes sense - esp given that a good chunk of this will be ripped out anyway, I'll just reintroduce it as regular values to minimize diff

File: pkg/config/config.go
Comment (jaiveerk) [reply pkg/config/config.go:43]: documenting our discussion here - agreed that resource parsing can happen on the add-on side, and we don't need to fill everything in via individual flags. can instead parse a lot of that from resource ID that we can feed in as a flag. Refactoring entire flag system will be a pretty big PR here and on RP side that is out of scope for this feature, but will jump on that in a new set of PRs following this one.

File: devenv/scripts/deploy_private_e2e_tester.sh
Comment (davidgamero) [inline devenv/scripts/deploy_private_e2e_tester.sh:17]: commenting the prefix/suffix removal would make this clearer for anyone who isn't as famiar with bash

File: devenv/scripts/deploy_private_e2e_tester.sh
Comment (davidgamero) [inline devenv/scripts/deploy_private_e2e_tester.sh:39]: nit: several blank lines here

File: devenv/main.tf
Comment (jaiveerk) [reply devenv/main.tf:330]: yep good catch!

File: devenv/scripts/push_addon_image.sh
Comment (jaiveerk) [reply devenv/scripts/push_addon_image.sh:3]: good idea! this will definitely make debugging easier

**Review by davidgamero — APPROVED** [review]
lgtm

File: pkg/manifests/manifests.go
Comment (jveski) [inline pkg/manifests/manifests.go:100]: I'd prefer to strategically set particular flags in the existing DNS deployment rather than forking the args entirely here.

File: pkg/manifests/manifests.go
Comment (jveski) [inline pkg/manifests/manifests.go:30]: nit: I personally find it more expressive to keep the kinds in struct string literals since they're commonly seen as literals in yaml manifests. Wouldn't block on it so do what you will. :)

Dig the other consts though - no more digging through the manifest to find the image tag.

File: pkg/manifests/manifests.go
Comment (jveski) [inline pkg/manifests/manifests.go:243]: Would prefer to merge this logic into `newIngressControllerService` to avoid adding another "hop" for future readers

File: pkg/manifests/manifests.go
Comment (jveski) [inline pkg/manifests/manifests.go:251]: Update specific keys instead of replacing the entire map pls

File: e2e/e2e_test.go
Comment (jveski) [inline e2e/e2e_test.go:30]: ha thanks for correcting 🙃

File: devenv/scripts/deploy_addon_private_cluster.sh
Comment (jveski) [inline devenv/scripts/deploy_addon_private_cluster.sh:2]: Can we manage a local env file with Terraform that contains the exports? Best to avoid parsing the tfstate since it isn't exactly a stable versioned api

File: e2e/Dockerfile
Comment (jveski) [inline e2e/Dockerfile:1]: What do you think about running the non-private tests containerized like this as well? Even though it might be a bit more brittle in general, seems like standardizing on one approach for private/non-private is a good enough justification.

File: devenv/scripts/make_e2e.sh
Comment (jveski) [inline devenv/scripts/make_e2e.sh:2]: is this file needed?

File: devenv/scripts/push_addon_image.sh
Comment (jveski) [inline devenv/scripts/push_addon_image.sh:1]: should this script be named `push_image`? From this project's perspective it isn't really an addon

**Review by jveski — CHANGES_REQUESTED** [review]

File: e2e/Dockerfile
Comment (OliverMKing) [reply e2e/Dockerfile:1]: I think this is the right way as well. It makes it so the point of these e2e tests are solely to test the WAR code while rp tests tests ensure the code there.

File: devenv/scripts/make_e2e.sh
Comment (jaiveerk) [reply devenv/scripts/make_e2e.sh:2]: I couldn't get the docker image to run the command without calling it from within a shell script. I can try to make it work without it if you'd like, but I spent a while and kept getting a weird "exec" error

File: pkg/manifests/manifests.go
Comment (jaiveerk) [reply pkg/manifests/manifests.go:100]: I'm not sure what you mean. Do you mean we should have this check (and resulting args) be determined within the newExternalDNSDeployment function? Or do you mean something along the lines of 
- assign initial args
- check if private
- modify individual entries/add necessary entries if so

File: pkg/manifests/manifests.go
Comment (jaiveerk) [reply pkg/manifests/manifests.go:30]: we're changing this a good deal with @OliverMKing 's next PR anyway, but I've personally run into issues when my string for the kind of resource wasn't exactly what it needs to be, so I wanted to keep a few of them as consts.

File: devenv/scripts/deploy_addon_private_cluster.sh
Comment (jaiveerk) [reply devenv/scripts/deploy_addon_private_cluster.sh:2]: do you mean we should be writing those values to an external file/reading them from there rather than from the tfstate?

File: e2e/Dockerfile
Comment (jaiveerk) [reply e2e/Dockerfile:1]: I don't see why not! Can definitely look into doing this

File: devenv/scripts/push_addon_image.sh
Comment (jaiveerk) [reply devenv/scripts/push_addon_image.sh:1]: fair point!

File: devenv/main.tf
Comment (bfoley13) [inline devenv/main.tf:10]: With more additions to config and complex feature I think this wont be viable. I think it would be better to have separation in the tf files used for specific tests/dev than clump it all into one place

File: devenv/scripts/push_addon_image.sh
Comment (bfoley13) [inline devenv/scripts/push_addon_image.sh:1]: Do you think we could split this PR into one thats changing the current E2E testing and then a follow up on enabling private clusters?

**Review by bfoley13 — CHANGES_REQUESTED** [review]

File: devenv/main.tf
Comment (jaiveerk) [reply devenv/main.tf:10]: with the number of repeated resources across both test types I wasn't sure, but are you basically saying we should have one TF file for private clusters and one for public?

File: devenv/scripts/push_addon_image.sh
Comment (jaiveerk) [reply devenv/scripts/push_addon_image.sh:1]: not exactly. combining both regular and private zone e2e testing relies on making the public suite use specific functionality (namely, the `az aks command invoke` command for deploying/monitoring everything). I could in theory try merging one PR for public stuff and then just rewrite whatever needs to be rewritten to make it compatible with private, but that doesn't seem like the smartest way to move forward imo.

File: pkg/manifests/manifests.go
Comment (jaiveerk) [reply pkg/manifests/manifests.go:251]: prior to my transformation the serviceObj doesn't have a map at serviceObj.ObjectMeta.Annotations - there's only a name, namespace, and labels

File: pkg/manifests/manifests.go
Comment (jveski) [reply pkg/manifests/manifests.go:100]: > Do you mean we should have this check (and resulting args) be determined within the newExternalDNSDeployment function?

Yes. 🙃

File: pkg/manifests/manifests.go
Comment (jaiveerk) [reply pkg/manifests/manifests.go:251]: following discussion, I will add a blank map in the original annotations position.

File: pkg/manifests/manifests.go
Comment (jveski) [inline pkg/manifests/manifests.go:477]: Can we DRY out the flags a bit?

Something like this:


```go
// Only set shared values in the struct
						Args: []string{
							"--source=ingress",
							"--interval=3m0s",
							"--txt-owner-id=" + conf.DNSRecordID,
							"--domain-filter=" + conf.DNSZoneDomain,
						},

// Then append relevant flags using logic
var additionalArgs []string
if conf.DNSZonePrivate {
		additionalArgs = []string{
			fmt.Sprintf("--provider=%s", azurePrivateDNSProvider),
			fmt.Sprintf("--azure-subscription-id=%s", conf.DNSZoneSub),
		}
} else {
		additionalArgs = []string{
			"--provider=azure",
		}
}
deploymentObj.Spec.Template.Spec.Containers[0].Args = append(deploymentObj.Spec.Template.Spec.Containers[0].Args, additionalArgs...)
```

File: pkg/manifests/manifests.go
Comment (jveski) [inline pkg/manifests/manifests.go:255]: 👍

File: devenv/main.tf
Comment (jveski) [inline devenv/main.tf:361]: Do you think it would make sense to fold the values of the two new local_file resources into the old e2econf?

I can see some value in keeping unrelated values in separate files, but at first glance there doesn't seem to be a conceptual difference between the various local files.

Comment (jaiveerk) [reply devenv/main.tf:361]: there's no conceptual difference but it has more to do with where the information is needed, namely that the e2e conf gets JSON-marshalled into a struct so I wanted to keep that to the values that are used within the testing suite itself

File: pkg/manifests/manifests.go
Comment (jaiveerk) [reply pkg/manifests/manifests.go:477]: yep that totally makes sense!

File: devenv/main.tf
Comment (jveski) [reply devenv/main.tf:361]: That makes sense. I would personally jam them all into one file so future consumers of the values don't need to worry about which file to read. But I'm cool either way

Comment (jaiveerk) [reply devenv/main.tf:286]: plus I think the conceptual separation makes sense given that the cluster_info isn't used by the test suite but rather the scripts that deploy the test, and that the other local file is just used by private tests (pushing a commit soon that separates the tf files though and that'll demonstrate that last piece pretty clearly)

File: Makefile
Comment (OliverMKing) [inline Makefile:4]: would be good to just have a `clean` that does both as well

File: Makefile
Comment (OliverMKing) [inline Makefile:40]: should this depend on the deploy-e2e and push-tester-image ones? I'd like to be able to run make e2e or make run-e2e to run an e2e test. Right now it looks like I'd have to run

make deploy-e2e then make run-e2e

File: Makefile
Comment (OliverMKing) [inline Makefile:20]: the way we actually use this is becoming a lot more complicated. comments or docs explaining how to use these would be good.

this is becoming a lot more complicated than just 
make dev
make push
make e2e

**Review by OliverMKing — CHANGES_REQUESTED** [review]

File: Makefile
Comment (jaiveerk) [reply Makefile:4]: makes sense!

File: Makefile
Comment (jaiveerk) [reply Makefile:20]: yep good call! I'll add a README

File: Makefile
Comment (jaiveerk) [reply Makefile:40]: you don't have to run make run-e2e. that gets run by the containerized tester. you would just run deploy-e2e. I'll document this in the README I mentioned in my reply to the other comment, I can definitely see that this is pretty confusing with no other context.

File: devenv/scripts/push_addon_image.sh
Comment (jaiveerk) [reply devenv/scripts/push_addon_image.sh]: update: after discussion, switching to this approach of merging public e2e changes to minimize size of this PR

File: README.md
Comment (OliverMKing) [inline README.md:29]: should this say private?

File: README.md
Comment (OliverMKing) [inline README.md:30]: this too seems like it should say private

Comment (jaiveerk) [reply README.md:30]: it absolutely should - noticed the typo, but separating this into two PRs, one for updating the public e2e tests/e2e framework, and then a second one for all things private DNS. Will look at the diff once the other is merged in and take care of all changes that need to get done.

**Pull request closed without merging**
